design/open/staging-compaction.md

```diff
@@ -0,0 +1,172 @@
+# Proposal: Staging Compaction
+
+## Problem Description
+
+lakeFS staging area is built on top of the [KV store](../accepted/metadata_kv/index.md).
+In short, the structure for the `Branch` entity in the kv-store is as follows:
+```go
+type Branch struct {
+	CommitID     CommitID
+	StagingToken StagingToken
+	// SealedTokens - Staging tokens are appended to the front, this allows building the diff iterator easily
+	SealedTokens []StagingToken
+}
+```
+
+Uncommitted entries are read from the staging token first, then from the 
+sealed tokens by order. Writes are performed on the staging token.
+
+The KV design has proven to meet lakeFS requirements for its consistency 
+guarantees. For most use-cases of uncommitted areas in lakeFS, it does that 
+efficiently. However, there are some cases where it falls short, for example 
+this [issue](https://github.com/treeverse/lakeFS/issues/2092).  
+There might be other cases where the structure of the `Branch` entity impacts 
+the performance of reading from the staging area, for example when the 
+number of sealed tokens is large (N)[^1] and reading a missing entry requires 
+reading from all N sealed tokens. 
+
+## Goals
+
+- Preserve lakeFS consistency guarantees for all scenarios.
+- A more efficient way to read from branches with a large number of 
+  tombstones (i.e. Fix the issue mentioned above).
+- Preserve lakeFS performance for other scenarios.[^2]
+
+
+## Non Goals
+
+- Improve the performance of the KV store in general.
+- Simplify the uncommitted area model in lakeFS. While we'll try to keep the 
+  changes minimal, we might need to add some complexity to the model to 
+  achieve the goals.
+
+
+## Proposed Design
+
+We propose to add a compaction mechanism to the staging area. We'll explain 
+how to decide when to compact (Sensor), how to compact (Compactor), 
+data read/write flows and commit flows after compaction.
+
+```go
+type Branch struct {
+	CommitID     CommitID
+	StagingToken StagingToken
+	// SealedTokens - Staging tokens are appended to the front, this allows building the diff iterator easily
+	SealedTokens []StagingToken
+	
+	CompactedMetaRange MetaRangeID
+	CompactionStartTime time.Time
+}
+```
+
+### Sensor
+
+The Sensor is responsible for deciding when to compact the staging area.
+The Sensor will be linked to the Graveler and will collect 
+information on writes to the staging area. It will decide when to compact a 
+certain branch based on the number of deleted entries to its staging area, 
+taking into accounts operations like commits/resets etc. It can also decide 
+based on the number of sealed tokens, although not necessarily a priority 
+for the first version. We can probably avoid the need to query the kv-store 
+to retrieve that information (except for the service startups) by caching 
+the information in the Sensor.
```

Comment:
It is hard to cache the number of deletions locally in a cluster of lakeFSes.  One option might be for each lakeFS instance to select a unique ID and store its deletions count on KV some key.  Writes are immediately uncontended, and write load can be reduced by batching the increment for a short while.